mage_ai/io/cratedb.py

```diff
@@ -0,0 +1,389 @@
+        
+        # force USE INSERT
+        use_insert_command = True
+
+        def clean_array_value(val):
+            if val is None or type(val) is not str or len(val) < 2:
+                return val
+            if val[0] == '[' and val[-1] == ']':
+                return '{' + val[1:-1] + '}'
+            return val
+
+        def serialize_obj(val):
+            if type(val) is dict or type(val) is np.ndarray:
+                return simplejson.dumps(
+                    val,
+                    default=encode_complex,
+                    ignore_nan=True,
+                )
+            elif type(val) is list and len(val) >= 1 and type(val[0]) is dict:
+                return simplejson.dumps(
+                    val,
+                    default=encode_complex,
+                    ignore_nan=True,
+                )
+            elif not use_insert_command and type(val) is list:
+                return clean_array_value(simplejson.dumps(
+                    val,
+                    default=encode_complex,
+                    ignore_nan=True,
+                ))
+            return val
+
+        df_ = df.copy()
+        columns = df_.columns
+
+        for col in columns:
+            df_col_dropna = df_[col].dropna()
+            if df_col_dropna.count() == 0:
+                continue
+            if dtypes[col] == PandasTypes.OBJECT \
+                    or (df_[col].dtype == PandasTypes.OBJECT and not
+                        isinstance(df_col_dropna.iloc[0], str)):
+                df_[col] = df_[col].apply(lambda x: serialize_obj(x))
+        df_.replace({np.NaN: None}, inplace=True)
+
+        insert_columns = ', '.join([f'"{col}"'for col in columns])
+
+        if use_insert_command:
+            # Use INSERT command
+            values_placeholder = ', '.join(["%s" for i in range(len(columns))])
+            values = []
+            for _, row in df_.iterrows():
+                values.append(tuple(row))
+            commands = [
+                f'INSERT INTO {full_table_name} ({insert_columns})',
+                f'VALUES ({values_placeholder})',
+            ]
+
+            unique_constraints = \
+                [f'"{self._clean_column_name(col, allow_reserved_words=allow_reserved_words)}"'
+                 for col in unique_constraints]
+            columns_cleaned = \
+                [f'"{self._clean_column_name(col, allow_reserved_words=allow_reserved_words)}"'
+                 for col in columns]
+
+            commands.append(f"ON CONFLICT ({', '.join(unique_constraints)})")
+            if UNIQUE_CONFLICT_METHOD_UPDATE == unique_conflict_method:
+                update_command = [f'{col} = EXCLUDED.{col}' for col in columns_cleaned]
+                commands.append(
+                    f"DO UPDATE SET {', '.join(update_command)}",
+                )
+            else:
+                commands.append('DO NOTHING')
+            cursor.executemany('\n'.join(commands), values)
```

Comment:
The `bulk_parameters` argument may want to be used here when applicable, depending on context.
```suggestion
            cursor.execute('\n'.join(commands), bulk_parameters=values)
```

Reply:
Forget about this comment. If it's psycopg, then there isn't such an argument.